persist/fs/retriever.go

```diff
@@ -242,16 +261,33 @@ func (r *blockRetriever) fetchBatch(
 	// Sort the requests by offset into the file before seeking
 	// to ensure all seeks are in ascending order
 	for _, req := range reqs {
-		req.seekOffset = seeker.SeekOffset(req.id)
+		entry, err := seeker.SeekIndexEntry(req.id)
+		if err != nil && err != errSeekIDNotFound {
+			req.onError(err)
+			continue
+		}
+
+		if err == errSeekIDNotFound {
+			req.doesNotExist = true
+		}
+		req.indexEntry = entry
 	}
 	sort.Sort(retrieveRequestByOffsetAsc(reqs))
 
 	// Seek and execute all requests
 	for _, req := range reqs {
-		data, err := seeker.Seek(req.id)
-		if err != nil && err != errSeekIDNotFound {
-			req.onError(err)
-			continue
+		var data checked.Bytes
+		var err error
+		// If the ID does not exist, don't try to seek to it or we'll get a checksum
+		// mismatch error because default offset value for indexEntry is zero.
+		if req.doesNotExist {
+			data, err = nil, nil
+		} else {
+			data, err = seeker.SeekByIndexEntry(req.indexEntry)
+			if err != nil && err != errSeekIDNotFound {
+				req.onError(err)
+				continue
+			}
```

Comment:
How about just `if !req.doesNotExist { ... }` then you don't have to set data and err to nil explicitly?

Reply:
good call